lib/Sema/ConstraintSystem.h

```diff
@@ -4643,6 +4643,69 @@ static inline bool computeTupleShuffle(TupleType *fromTuple,
 /// TinyPtrVector for unsigned values.
 using ParamBinding = SmallVector<unsigned, 1>;
 
+/// It holds all interested computation result of \c matchCallArguments
+struct CallArgumentMatchResult {
```

Comment:
Note that we'd like to allocate as little as possible in a "happy" case, so I think the result of `matchCallArguments` should stay the same (array of arrays with positions). We should track all the information in `ArgumentFailureTracker` listener when it's relevant. Which brings us to the work performed issue, I think until solver is in "diagnostic" mode we shouldn't try to fix or track anything and do as little as possible.

It might make sense to just scrap current implementation and start from scratch having in mind all the things it needs to support regarding diagnostics. Maybe have it encapsulated into a struct/class so it could be re-used. 

Another  issue is `diagnoseArgumentLabelError` - I think that shouldn't exist because it's too late to diagnose that post-factum when solution is already applied. Currently I see it being used only in `CSDiagnostics` and `TypeCheckAvailability` we need to take a closer look at the second use-case and see whether it even makes sense or not. All of the logic currently associated with `diagnoseArgumentLabelError` should be a part of `matchCallArguments`.

Reply:
Thank you for your comment.

I understood that keeping memory low is important in "happy" case.

I finished reworking from scratch just now.
It doesn't use `diagnoseArgumentLabelError` already.